contributors/design-proposals/device-plugin.md

```diff
@@ -0,0 +1,410 @@
+            labels:
+                - name: device-plugin
+        spec:
+            containers:
+                name: device-plugin-ctr
+                image: nvidia/device-plugin:1.0
+                volumeMounts:
+                  - mountPath: /device-plugin
+                  - name: device-plugin
+           volumes:
+             - name: device-plugin
+               hostPath:
+                   path: /var/run/kubernetes
+```
+
+## API Changes
+### Device
+
+When discovering the devices, Kubelet will be in charge of advertising those
+resources to the API server.
+
+We will advertise each device returned by the Device Plugin in a new structure
+called `Device`.
+It is defined as follows:
+
+```golang
+type Device struct {
+	Kind       string
+	Vendor     string
+	Name       string
+	Health     DeviceHealthStatus
+	Properties map[string]string
+}
+```
+
+Because the current API (Capacity) can not be extended to support Device,
+we will need to create two new attributes in the NodeStatus structure:
+  * `DevCapacity`: Describing the device capacity of the node
+  * `DevAvailable`: Describing the available devices
+
+```golang
+type NodeStatus struct {
+	DevCapacity []Device
+	DevAvailable []Device
+}
+```
+
+We also introduce the `Allocated` field in the pod's status so that user
+can know what devices were assigned to the pod. It could also be useful in
+the case of monitoring
+
+```golang
+type ContainerStatus struct {
+	Devices []Device
+}
+```
+
+## Device Plugin
+
+The device plugin is structured in 5 parts:
+1. Registration: The device plugin advertises it's presence to Kubelet
+2. Discovery: Kubelet calls the device plugin to list it's devices
+3. Allocate / Deallocate: When creating/deleting containers requesting the
+   devices advertised by the device plugin, Kubelet calls the device plugin's
+   `Allocate` and `Deallocate` functions.
+4. Cleanup: Kubelet terminates the communication through a "Stop"
+4. Heartbeat: The device plugin polls Kubelet to know if it's still alive
+   and if it has to re-issue a Register request
+
+### Registration
+
+When starting the device plugin is expected to make a (client) gRPC call
+to the `Register` function that Kubelet exposes.
+
+The communication between Kubelet is expected to happen only through Unix
+sockets and follow this simple pattern:
+1. The device plugins starts it's gRPC server
+2. The device plugins sends a `RegisterRequest` to Kubelet (through a
+   gRPC request)
+3. Kubelet calls the `Init` function exposed by the device plugin
+4. Kubelet starts it's Discovery phase and calls `Init`, `Discover`, `Monitor`
+5. Kubelet answers to the `RegisterRequest` with a `RegisterResponse`
+   containing any error Kubelet might have encountered
+
+### Unix Socket
+
+Device Plugins are expected to communicate with Kubelet through gRPC
+on an Unix socket.
+When starting the gRPC server, they are expected to create a unix socket
+at the following host path: `/var/run/kubernetes`.
```

Comment:
Device plugins can share dynamic port at register phase, but to start register phase device plugins will need kubelet server to listen either on a static(configured) or named(kubelet.sock) port. If we choose static port for kubelet, then i am able to see how this can help reduce the knobs outside the API. But we choose to go with a named port for kubelet server, in that case i am unable to see how are reducing knobs by using dynamic ports for device plugins.